### PR TITLE
REST API test improvements

### PR DESCRIPTION
**Pull request by atzurk**
#### Description
This PR includes  the following:

Added tests to automatically verify all the implemented functionalities and data fields from the Patient and User entities.

Added function to verify LocalDate objects

### CODE REVIEW

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 1]
```diff
@@ -67,10 +85,23 @@ private void testCreateInvalidPatient(Patient patient) {
...
+        testCreateInvalidPatient(getPatientData("", "Jameson", "jj@.", "333"));
+        testCreateInvalidPatient(getPatientData("Thomas", "", "", ""));
+        testCreateInvalidPatient(getPatientData("Thomas", "", "", "111"));
+        testCreateInvalidPatient(getPatientData("Thomas", "", "jj@example.com", ""));
+        testCreateInvalidPatient(getPatientData("John", "", "jburrows@test,com", "222"));
+        testCreateInvalidPatient(getPatientData("Thomas", "Summers", "", ""));
+        testCreateInvalidPatient(getPatientData("Thomas", "Summers", "", "333"));
+        testCreateInvalidPatient(getPatientData("Thomas", "Summers", "tsms@example,com", ""));
+        testCreateInvalidPatient(getPatientData("John", "Burrows", "jburrows,", "345423"));
+        //TODO: All possible combinations and patterns(email)
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: I also think that it's not required to test all possible combinations. We just need to test that each field is null at least once.
We need to add a step to test the email incorrect format.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 2]
```diff
@@ -106,6 +137,7 @@ public void testGetPatientList() {
         long initialSize = patientsBefore.getTotalElements();
         savePatientsDatabase(patientsBefore.getContent());
         TestUtil.checkANumber("List size", numberOfPatients, initialSize);
+        // TODO: Verify the elements in the list also.
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: Remove the TODO: comment.

Comment (olgapuig) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: I would not document any expected result for the method createSomePatients as it is an auxiliar method

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 3]
```diff
@@ -127,6 +148,7 @@ public void testGetPatient() {
                 .then().assertThat().statusCode(200)
                 .extract().as(Patient.class);
         //I would assign ID and use the value to report in the exp.result
+        // TODO: Get rid of excessive documentation (some verifications shouldn't be reported)
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: Remove the TODO: comments

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 4]
```diff
@@ -127,6 +148,7 @@ public void testGetPatient() {
...
         TestUtil.checkObjectIsNotNull("Patient ID " + patientCreated.getId(), patientCreated.getId());
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: I would use just an assert without futher documentation.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 5]
```diff
@@ -181,4 +206,6 @@ public void testDeleteUnexistingPatient() {
                 .extract().statusCode();
         TestUtil.checkField("Status Code", 404, statusCode);
     }
+
+    // TODO: Include update patient,
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: Remove TODO:

File: src/test/java/com/systelab/seed/user/UserResourceTest.java [hunk 6]
```diff
@@ -30,6 +30,7 @@ public void testGetUserList() {
         TestUtil.checkObjectIsNotNull("Users", users);
     }
 
+    // TODO: Pending login verification
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/user/UserResourceTest.java]: Remove TODO:

File: src/test/java/com/systelab/seed/user/UserResourceTest.java [hunk 7]
```diff
@@ -30,6 +30,7 @@ public void testGetUserList() {
...
     @Description("Create a User with name, login and password")
     @Test
     public void testCreateUser() {
```
Comment (olgapuig) [inline src/test/java/com/systelab/seed/user/UserResourceTest.java]: I would create additional scenario to test Create User without username, password and name.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 8]
```diff
@@ -41,10 +44,19 @@ private Patient getPatientData(String name, String surname, String email) {
         return patient;
     }
 
-    @Description("Create a new patient with name, surname and email")
+    private Patient getPatientData(String name, String surname, String email, String medicalNumber) {
```
Comment (danipiques) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: The getPatientData with 4 arguments should replace the previous getPatientData with only 3. Additionally I would add the parameters that are not being used DOB, Street, City, Zip Code

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 1 above)
Comment (danipiques) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: I agree. In an automated test we can add as cases as we want, but in this case, as a seed example, it should be better to left only one case of each mandatory field. Suggestion:
```suggestion
        testCreateInvalidPatient(getPatientData("", "", "", ""));
        testCreateInvalidPatient(getPatientData("", "Summers", "jsummers@email.com", "345423"));
        testCreateInvalidPatient(getPatientData("John", "", "jsummers@email.com", "345423"));
        testCreateInvalidPatient(getPatientData("John", "Summers", "", "345423"));
        testCreateInvalidPatient(getPatientData("John", "Summers", "jsummers@email.com", ""));
```

Comment (danipiques) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: To get a test clearer, the cases of email format wrong (and other invalid patient reasons), should be added in other steps, and the Action should indicate clearly what is the reason of the invalid data (the expected result only indicates the error code). For example:

@Description("Create a Patient with invalid data: mandatory fields empty: Name, Surname, DOB or email")
public void testCreatePatientWithMandatoryFieldsEmpty() {
    testCreateInvalidPatient...
}

@Description("Create a Patient with invalid data: email format")
public void testCreatePatientWithEmailFormatInvalid() {
    testCreateInvalidPatient...
}

@Description("Create a Patient with invalid data: unique fields: Medical Number")
public void testCreatePatientWithUniqueFieldsRepeated() {
    testCreateInvalidPatient...
}

Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: In the Patient entity only 'Name' and 'Surname' have the Non-Null constraint. Additionally, the email format is verified so I will include some tests with invalid emails.

The unique field is not defined in the Patient entity, so it's no point in verify that right now.

I think this two steps are more close to what is defined by the entity:
@Description("Create a Patient with invalid data: mandatory fields empty (name, surname)")
@Description("Create a Patient with invalid data: email format")

File: src/test/java/com/systelab/seed/user/UserResourceTest.java (hunk 7 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/user/UserResourceTest.java]: done

File: src/test/java/com/systelab/seed/user/UserResourceTest.java (hunk 6 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/user/UserResourceTest.java]: done

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 5 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: done

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 3 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: done

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 4 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: done

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 8 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: done

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 9]
```diff
@@ -118,8 +157,8 @@ private void createSomePatients(int numberOfPatients) {
                     .when().post("/patients/patient")
                     .then().assertThat().statusCode(200)
                     .extract().as(Patient.class);
-            //This check is identical for all patients and shall be specific or merged in one informing about creation of {0} patients.
-            TestUtil.checkObjectIsNotNull("Patient ID " + patientCreated.getId(), patientCreated.getId());
+
+            Assertions.assertNotNull(patientCreated.getId(), "Patient not created");
```
Comment (ManelMoreno) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: The intention of the comment you just deleted was to have an specific expected result for each patient created (using ID or something shorter like patient 1, patient 2,...patient n, OR use a unique expected result for the whole batch of patients created like "5 patients successfully created".
Now, I can not understand your "Patient not created" message.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 10]
```diff
@@ -194,7 +232,7 @@ public void testDeletePatient() {
                 .then().assertThat().statusCode(200)
                 .extract().as(Patient.class);
 
-        Assertions.assertNotNull(patientCreated);
+        Assertions.assertNotNull(patientCreated, "Patient not created");
```
Comment (ManelMoreno) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: Do you mean "Patient created"? If it is not created, you can not delete it and test the get by non-existing id...

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java [hunk 11]
```diff
@@ -53,24 +52,67 @@ private Patient getPatientData(String name, String surname, String email, String
...
+        TestUtil.checkField("Medical Number", expectedMedNumber, patientCreated.getMedicalNumber());
+        TestUtil.checkField("DOB", expectedDob, patientCreated.getDob());
+        TestUtil.checkField("Street", expectedStreet, patientCreated.getAddress().getStreet());
+        TestUtil.checkField("City", expectedCity, patientCreated.getAddress().getCity());
+        TestUtil.checkField("Zip", expectedZip, patientCreated.getAddress().getZip());
+    }
+
+    @Description("Create a new patient with name, surname, email, medical number, dob, and complete address")
+    @Test
+    public void testCreatePatientWithAllInfo() {
```
Comment (ManelMoreno) [inline src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: We could also create valid and invalid Patients, using boundary field values, as example of boundary values examples.
(This is supposing that current fields implement some valid ranges that we could test..)

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 9 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: It was agreed with Olga that some verifications or checks in the returned values of auxiliar functions would not be reported to allure (as they are reported in other tests), so I replaced the TestUtil.checkObjectIsNotNull for the Assertions.assertNotNull. The verification is equally done in both cases. But the Assertion does not report anything to allure, and just displays the 'Patient not created' message when the assertion fails, that is, when the patients has not been created, and the deletion cannot be verified.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 10 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: This mesage only is displayed when the assertion has failed, so the patient has not been created.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 11 above)
Comment (atzurk) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: Provided that the REST API is well documented and reflects the complete specification of the protocol we want to verify. Totally agree with that.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 10 above)
Comment (ManelMoreno) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: I'm sorry, I didn't catch that the string is just if the assert fails, so when patient is not created.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 9 above)
Comment (olgapuig) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: "Patient not created" is the message that will be displayed in case the assertion is not null

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 10 above)
Comment (olgapuig) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: "Patient not created" is the message that will be displayed in case the assertion is not null

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 11 above)
Comment (olgapuig) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: I agree on this.

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 9 above)
Comment (ManelMoreno) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: Ok, got it. The declaration in TestUtils confused me (even worse, this new assertion does not use TestUtils..).

File: src/test/java/com/systelab/seed/patient/PatientResourceTest.java (hunk 11 above)
Comment (alfonsserra) [reply src/test/java/com/systelab/seed/patient/PatientResourceTest.java]: > We could also create valid and invalid Patients, using boundary field values, as example of boundary values examples.
> (This is supposing that current fields implement some valid ranges that we could test..)

There is an specific method for that (testCreateInvalidPatient())

File: src/test/java/com/systelab/seed/utils/TestUtil.java [hunk 12]
```diff
@@ -55,6 +56,9 @@ public static void checkField(String field, Boolean value, Boolean returnedValue
         Assertions.assertEquals(value, returnedValue);
     }
 
+    @Step("Field {0} is equal to {1}")
+    public static void checkField(String field, LocalDate value, LocalDate returnedValue) {Assertions.assertEquals(value, returnedValue); }
+
     @Step("Field {0} is equal to {1} Â± {3}")
```
Comment (alfonsserra) [inline src/test/java/com/systelab/seed/utils/TestUtil.java]: Be careful with the Â± character

Comment (atzurk) [reply src/test/java/com/systelab/seed/utils/TestUtil.java]: fixed.
